src/dataflow/src/render/join/mod.rs

```diff
@@ -0,0 +1,275 @@
+/// may already have been partially subjected to logic we need to apply.
+pub struct JoinBuildState {
+    /// Map from expected locations in extended output column reckoning to physical locations.
+    pub column_map: HashMap<usize, usize>,
+    /// A list of equivalence classes of expressions.
+    ///
+    /// Within each equivalence class, expressions must evaluate to the same result to pass
+    /// the join expression. Importantly, "the same" should be evaluated with `Datum`s Rust
+    /// equality, rather than the equality presented by the `BinaryFunc` equality operator.
+    /// The distinction is important for null handling, at the least.
+    pub equivalences: Vec<Vec<ScalarExpr>>,
+    /// The linear operator logic (maps, filters, and projection) that remains to be applied
+    /// to the output of the join.
+    ///
+    /// We we advance through the construction of the join dataflow, we may be able to peel
+    /// off some of this work, ideally reducing `mfp` to something nearly the identity.
+    pub mfp: MapFilterProject,
+}
+
+impl JoinBuildState {
+    /// Create a new join state and initial closure from initial values.
+    ///
+    /// The initial closure can be `None` which indicates that it is the identity operator.
+    pub fn new(
+        columns: std::ops::Range<usize>,
+        equivalences: &[Vec<ScalarExpr>],
+        mfp: &MapFilterProject,
+    ) -> Self {
+        let mut column_map = HashMap::new();
+        for column in columns {
+            column_map.insert(column, column_map.len());
+        }
+        Self {
+            column_map,
+            equivalences: equivalences.to_vec(),
+            mfp: mfp.clone(),
+        }
+    }
+
+    /// Present new columns and extract any newly available closure.
+    pub fn add_columns(
+        &mut self,
+        new_columns: std::ops::Range<usize>,
+        bound_expressions: &[ScalarExpr],
+    ) -> JoinClosure {
+        // Remove each element of `next_keys` from `equivalences`, so that we
+        // avoid redundant predicate work. This removal also paves the way for
+        // more precise "demand" information going forward.
+        for equivalence in self.equivalences.iter_mut() {
+            equivalence.retain(|expr| !bound_expressions.contains(expr));
+        }
+        self.equivalences.retain(|e| e.len() > 1);
+
+        // Update our map of the sources of each column in the update stream.
+        for column in new_columns {
+            self.column_map.insert(column, self.column_map.len());
+        }
+
+        self.extract_closure()
+    }
+
+    /// Extract a final `MapFilterProject` once all columns are available.
+    ///
+    /// The `None` output indicates that the `MapFilterProject` is the identity.
+    pub fn complete(self) -> Option<MapFilterProject> {
+        let Self {
+            column_map,
+            equivalences,
+            mut mfp,
+        } = self;
+        assert!(equivalences.is_empty());
```

Comment:
I haven't read the rest of the changes but this assert scares me a little bit -- at the very least maybe add a comment indicating when it is / isn't safe to call this function?

Reply:
Good point! What I could do is make it produce a `JoinClosure`, it just happens that the previous code didn't do that because it was pretty sure there were no equivalences. I think that is probably fine here, though (adds a minor bit of logic to check that there are no constraints, but .. makes the code less error-prone probably).